### PR TITLE
trivial: make symlinkJoin support pname+version alone

### PR DESCRIPTION
**Pull request by pbsds**
## Description of changes

inspired by https://github.com/Mic92/nix-update/issues/284

should be 0 rebuilds

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: doc/build-helpers/trivial-build-helpers.chapter.md
Comment (drupol) [inline doc/build-helpers/trivial-build-helpers.chapter.md:735]:
```suggestion
This can be used to put many derivations into the same directory structure. It works by creating a new derivation and adding symlinks to each of the paths listed. It expects two arguments, `name`, and `paths`. `name` (or alternatively `pname` and `version`) is the name used in the Nix store path for the created derivation. `paths` is a list of paths that will be symlinked. These paths can be to Nix store derivations or any other subdirectory contained within.
```

**Review by drupol — COMMENTED** [review]
LGTM but I'd wait for another review.

**Review by philiptaron — APPROVED** [review]
Yeah, this is fine. We could finesse it a little:

1. Better error message if nothing is specified.
2. Better error message if name AND pname || version is specified
3. Better error message if pname but not version is specified and vice versa
4. Don't pass through the pname or version into runCommand's derivationArgs (name already omitted)

**Review by getchoo — CHANGES_REQUESTED** [review]
+1 on the errors mention by Philip, but the removal of `pname` and `version` here is a requirement IMO

**Comment by pbsds** [comment]
> 1. Better error message if nothing is specified.

Done

> 2. Better error message if name AND pname || version is specified

This is not an error in mkDerivation

> 3. Better error message if pname but not version is specified and vice versa

Possible, but i'm not sure if i like it:

```nix
assert lib.assertMsg (args_ ? pname && args_ ? version)
  "symlinkJoin requires either a `name` OR `pname` and `version`, found: ${lib.concatStringsSep ", " (lib.filter (lib.flip builtins.elem ["name" "pname" "version"]) (lib.attrNames args_))}";
"${args_.pname}-${args_.version}"
```

> 4. Don't pass through the pname or version into runCommand's derivationArgs (name already omitted)

That would mean not setting `finalPackage.pname` and `finalPackage.version`, which is the current behaviour and the entire reason why you want to provide these args

**Comment by pbsds** [comment]
> +1 on the errors mention by Philip, but the removal of `pname` and `version` here is a requirement IMO

This confuses me, is there a negation missing or is there something i'm missing?

**Comment by philiptaron** [comment]
> Handling this request in full is possible, but I'm not sure if I like it.

Here's my attempt, designed to be placed after in the `let` block.

```
assert lib.assertMsg (args_ ? name -> !(args ? pname || args ? version) "symlinkJoin requires either `name` or `pname` and `version`; to fix this, remove `name` or both of `pname` and `version`"
assert lib.assertMsg (args_ ? pname == args_ ? version) "symlinkJoin requires `pname` and `version` to be specified together";
```

**Comment by philiptaron** [comment]
> This confuses me, is there a negation missing or is there something I'm missing?

No, I think I'm the source of the issue: I didn't fully understand that getting `pname` and especially `version` into the resulting derivationArgs was a key goal of this patch.

**Review by philiptaron — APPROVED** [review]

**Comment by pbsds** [comment]
It does seem to achieve the intended hardening, but the breakage is massive and I assume the eval times will increase, especially in nixos.

**Comment by pbsds** [comment]
Drafted until the way forward is decided. Do we (1) forge ahead and establish this new hardening convention and patch all violations, or (2) drop https://github.com/NixOS/nixpkgs/commit/42ab48f2f77149ab15ea3ec2d398f45e1ec2245c ?

**Comment by philiptaron** [comment]
> Drafted until the way forward is decided. Do we (1) forge ahead and establish this new hardening convention and patch all violations, or (2) drop [42ab48f](https://github.com/NixOS/nixpkgs/commit/42ab48f2f77149ab15ea3ec2d398f45e1ec2245c) ?

Note the performance result: https://github.com/NixOS/nixpkgs/pull/344645/checks?check_run_id=30790294683

For me, given that there's _any_ eval troubles, I'm more on the drop-and-defer-to-a-future-PR train -- a PR that means to tackle this relatively common idiom of name vs. pname/version, and that makes a stand about how it ought to be represented in nixpkgs.

**Comment by philiptaron** [comment]
I'm still on the drop-and-defer-to-a-future-PR train. 🚋

**Comment by pbsds** [comment]
oh wow it has been two weeks. I was supposed to revert the commit but forgot

**Comment by pbsds** [comment]
pr is stuck at "processing" after I pushed, might be the current outage https://www.githubstatus.com/incidents/myvz2tsj2dh8

**Comment by philiptaron** [comment]
I don't think the relevant commit was dropped.

**Comment by pbsds** [comment]
image

**Comment by philiptaron** [comment]
> image

Dang, I think my eyes have never processed that section of the header before. 👀 Thanks for entertaining my blindness. 😁

**Review by philiptaron — APPROVED** [review]

**Pull request merged**
